src/mngt/ocf_mngt_cache.c

```diff
@@ -1921,6 +1933,168 @@ struct ocf_pipeline_properties _ocf_mngt_cache_bind_pipeline_properties = {
 	},
 };
 
+static void _ocf_mngt_activate_wait_metadata_io_finish(void *priv)
+{
+	struct ocf_cache_attach_context *context = priv;
+
+	ocf_pipeline_next(context->pipeline);
+}
+
+static void _ocf_mngt_activate_wait_metadata_io(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_cache_attach_context *context = priv;
+	ocf_cache_t cache = context->cache;
+
+	ocf_refcnt_freeze(&cache->refcnt.metadata);
+	ocf_refcnt_register_zero_cb(&cache->refcnt.metadata,
+			_ocf_mngt_activate_wait_metadata_io_finish, context);
+}
+
+static void _ocf_mngt_activate_check_superblock_complete(void *priv, int error)
+{
+	struct ocf_cache_attach_context *context = priv;
+	ocf_cache_t cache = context->cache;
+	int result;
+
+	if (error)
+		OCF_PL_FINISH_RET(context->pipeline, error);
+
+	result = ocf_metadata_validate_superblock(cache->owner,
+			cache->conf_meta);
+	if (result)
+		OCF_PL_FINISH_RET(context->pipeline, result);
+
+	ocf_pipeline_next(context->pipeline);
+}
+
+static void _ocf_mngt_activate_check_superblock(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_cache_attach_context *context = priv;
+	ocf_cache_t cache = context->cache;
+
+	ocf_metadata_sb_crc_recovery(cache,
+			_ocf_mngt_activate_check_superblock_complete, context);
+}
+
+static void _ocf_mngt_activate_compare_superblock_end(
+		struct ocf_metadata_read_sb_ctx *sb_ctx)
+{
+	struct ocf_superblock_config *superblock = &sb_ctx->superblock;
+	struct ocf_cache_attach_context *context = sb_ctx->priv1;
+	ocf_cache_t cache = context->cache;
+	int result, diff;
+
+	result = env_memcmp(cache->conf_meta, sizeof(*cache->conf_meta),
+			superblock, sizeof(*superblock), &diff);
+	if (result)
+		OCF_PL_FINISH_RET(context->pipeline, result);
+
+	if (diff) {
+		ocf_cache_log(cache, log_err, "Superblock missmatch!\n");
+		OCF_PL_FINISH_RET(context->pipeline, -OCF_ERR_INVAL);
+	}
+
+	ocf_pipeline_next(context->pipeline);
+}
+
+static void _ocf_mngt_activate_compare_superblock(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_cache_attach_context *context = priv;
+	ocf_cache_t cache = context->cache;
+	int result;
+
+	result = ocf_metadata_read_sb(cache->owner, ocf_cache_get_volume(cache),
+			_ocf_mngt_activate_compare_superblock_end,
+			context, NULL);
+	if (result)
+		OCF_PL_FINISH_RET(pipeline, result);
+}
+
+static void _ocf_mngt_activate_init_instance(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_cache_attach_context *context = priv;
+
+	_ocf_mngt_load_init_instance_recovery(context);
+}
+
+static void _ocf_mngt_activate_swap_stop_pipeline(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_cache_attach_context *context = priv;
+	ocf_cache_t cache = context->cache;
+	ocf_pipeline_t stop_pipeline;
+	int result;
+
+	result = ocf_pipeline_create(&stop_pipeline, cache,
+			&ocf_mngt_cache_stop_pipeline_properties);
```

Comment:
If `ocf_pipeline_create()` fails, `ocf_mngt_cache_stop_passive_pipeline_properties`  isn't suitable to stop the cache at this point; i.e. it doesn't handle removing cores thus stop pipelines have to swapped earlier. Also the stop pipeline has be created in `_ocf_mngt_cache_activate()`

Reply:
This is not a problem as `_ocf_mngt_cache_activate_finish()` handles cleanup after pipeline failure, so we leave in the same state as we started (i.e. _passive_ state).